src/resources/ble_esp32.cc

```diff
@@ -82,15 +82,145 @@ class DiscoveredPeripheral : public DiscoveredPeripheralList::Element {
   uint8 event_type_;
 };
 
+static const word kInvalidToken = -1;
+
+/// A map from tokens to BleResources.
+///
+/// Since NimBLE runs on a separate thread we can have callbacks for resources
+/// that have already been deleted. As such, we can't use pointers to our
+/// objects as callback arguments. Instead we create tokens that we then
+/// convert back to the actual BleResources (assuming they are still alive).
+class TokenResourceMap {
+ public:
+  ~TokenResourceMap() {
+    if (capacity != -1) {
+      free(entries);
+    }
+  }
+
+  // Returns false if there was a malloc error.
+  bool add(BleResource* resource, word* result) {
+    if (!reserve_space()) return false;
+    word token = sequence_counter++;
+    *result = token;
+    TokenResourceEntry entry = {token, resource};
+    entries[length++] = entry;
+    return true;
+  }
+
+  bool reserve_space() {
```

Comment:
I think you should avoid these super long methods inside the class definitions. You're asking the compiler to inline them.

Reply:
done.